sources/org.osbuild.containers-storage

```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/python3
+"""Provide access to an image in the host system's container storage.
+
+This stage differs from other sources in that the source check's to see if
+the container is available in the host's container storage. This puts a requirement
+on the user to ensure that the container is copied into local storage before trying
+to build an image. The starts by reading the host's `/etc/containers/storage.conf`
+file and then using the config to check if the container has been imported.
+
+Buildhost commands used: `skopeo`.
+"""
+
+import hashlib
+import subprocess as sp
+import sys
+
+from osbuild import sources
+from osbuild.util import containers
+
+SCHEMA = """
+"additionalProperties": false,
+"definitions": {
+  "item": {
+    "description": "The container image to fetch indexed by the container image id",
+    "type": "object",
+    "additionalProperties": false,
+    "patternProperties": {
+      "sha256:[0-9a-f]{64}": {
+        "type": "object",
+        "additionalProperties": false,
+        "required": ["image"],
+        "properties": {
+          "image": {
+            "type": "object",
+            "additionalProperties": false,
+            "required": ["name", "digest"],
+            "properties": {
+              "name": {
+                "type": "string",
+                "description": "Name of the image (including registry)."
+              },
+              "digest": {
+                "type": "string",
+                "description": "Digest of image in registry.",
+                "pattern": "sha256:[0-9a-f]{64}"
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+},
+"properties": {
+  "items": {"$ref": "#/definitions/item"},
+  "digests": {"$ref": "#/definitions/item"}
+},
+"oneOf": [{
+  "required": ["items"]
+}, {
+  "required": ["digests"]
+}]
+"""
+
+
+class ContainersStorageSource(sources.SourceService):
+
+    content_type = "org.osbuild.containers"
+
+    storage_conf = None
+
+    def local_image_name(self, imagename):
+        """
+        Construct the full image name that references an image with a given checksum in the local storage.
+        """
+        conf = containers.get_host_storage(self.storage_conf)
+        driver = conf["storage"]["driver"]
+        graphroot = conf["storage"]["graphroot"]
+        runroot = conf["storage"]["runroot"]
+        # TODO: include storage options
+        return f"containers-storage:[{driver}@{graphroot}+{runroot}]{imagename}"
+
+    def fetch_one(self, checksum, desc) -> None:
+        return
+
+    def fetch_all(self, items) -> None:
+        return
+
+    def exists(self, checksum, desc) -> bool:
+        image_id = checksum
+        image = desc["image"]
+        imagename = image["name"]
+        digest = image["digest"]
+
+        source = self.local_image_name(imagename)
+        res = sp.run(["skopeo", "inspect", "--raw", "--config", source], check=False, capture_output=True)
```

Comment:
I wonder if we can avoid the `capture_output`here - AFAICT res.std{out,err} is not used here and by capturing the output we will "swallow" error that might help us/our users to diagnose any issues (we had a similar issue in https://github.com/osbuild/osbuild/pull/1571)

Or alternatively we could include `res.stderr` in the exception below to help diagnose issues that are not a missing image (like corrupted container storage or rare/unexpected things like this)

Reply:
The issue I'm having here is that we are using the `res.stdout` to calculate the hash... but `res.stdout` is never empty because `skopeo` prints some warnings :/ so I've just added the `res.stderr` to the error that we raise below this